include: improve compatibility with C++ apps.

nzmichaelh opened the pull request:
This patch lets a C++ application use more of Zephyr by adding guards
and changeing some constructs to the C++11 equivalent.

Changes include:

- Adding guards
- Switching to static_assert
- Switching to a template for ARRAY_SIZE as g++ doesn't have the
  builtin.
- Re-ordering designated initialisers to match the struct field order
  as G++ only supports simple designated initialisers.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh committed:
include: improve compatibility with C++ apps.

This patch lets a C++ application use more of Zephyr by adding guards
and changeing some constructs to the C++11 equivalent.

Changes include:

- Adding guards
- Switching to static_assert
- Switching to a template for ARRAY_SIZE as g++ doesn't have the
  builtin.
- Re-ordering designated initialisers to match the struct field order
  as G++ only supports simple designated initialisers.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
This is failing as the tree is unstable:

zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -Wl,--end-group -L"/opt/sdk/zephyr-sdk-0.9.2/sysroots/armv5-zephyr-eabi/usr/lib/arm-zephyr-eabi/6.2.0/armv7e-m" -L/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/nrf52840_pca10056/samples/subsys/usb/cdc_acm/test_x86/zephyr -lgcc -Wl,--print-memory-usage -nostartfiles -nodefaultlibs -nostdlib -static -no-pie -Wl,--fatal-warnings -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none && :
Memory region         Used Size  Region Size  %age Used
           FLASH:       67364 B         1 MB      6.42%
            SRAM:       16756 B       256 KB      6.39%
        IDT_LIST:         164 B         2 KB      8.01zephyr/libzephyr.a(usb_device.c.obj): In function `usb_transfer_work':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/usb/usb_device.c:1109: undefined reference to `usb_dc_ep_mps'

nzmichaelh commented:
Tree is now stable, yay!

nashif commented on ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h:
```diff
@@ -58,6 +58,10 @@
 #ifndef __UECC_PLATFORM_SPECIFIC_H_
 #define __UECC_PLATFORM_SPECIFIC_H_
 
+#ifdef __cplusplus
```
please split this into its own commit and send tinycrypt patches to https://github.com/01org/tinycrypt/pulls

nashif commented on ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h:
```diff
@@ -78,4 +82,8 @@
 
 int default_CSPRNG(uint8_t *dest, unsigned int size);
 
+#ifdef __cplusplus
```
ditto

tyrantlizard77 reviewed (dismissed):
LGTM but please split out and upstream the bits that touch tinycrypt.

nzmichaelh committed:
ztest: define test_main in the header file.

This also makes test_main C++ compatible by setting the linkage.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh committed:
tests: add a C++ compile test.

The test case checks that many of the common header files are C++
compatible and checks that some of the primitives like ARRAY_SIZE
work.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
@andrewboie @nashif  I've split out the ext/ changes.

I also went through the most common driver files and added guards, and added a mainly compile related test to see that they stay compatible in the future.

mbolivar commented on tests/application_development/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
+
+FILE(GLOB app_sources src/*.cpp)
```
I think @SebastianBoe prefers not using GLOB in Zephyr's CMake files.

nashif replied:
git grep "GLOB app_sources" | wc -l
261

mbolivar replied:
Cool, never mind :+1:

mbolivar reviewed (commented):
led_strip changes are fine. One potential issue on the test CMakeLists.txt.

nzmichaelh commented:
I've updated spi_legacy.h as well to fix the compile test.

nzmichaelh committed:
cpp: mark __dso_handle as weak.

The POSIX target uses the native crt0 which already provides
__dso_handle.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh commented:
@nashif  Hi Anas.  Could you install lib32stdc++-7-dev on the Shippable configuration?  It's needed for C++ apps to link in the POSIX m32 config.

nzmichaelh commented:
PR sent for adding the m32 version of libsdc++ to the Shippable config.

nashif commented:
recheck

nashif commented on tests/application_development/cpp/src/main.cpp:
```diff
@@ -0,0 +1,51 @@
+	return 0;
+}
+
+SYS_INIT(test_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
why are you starting the test function this way and not via test_main?

nzmichaelh replied:
Clarified that this is to test SYS_INIT itself.

nzmichaelh commented:
recheck

nashif commented:
recheck

nashif commented on tests/application_development/cpp/testcase.yaml:
```diff
@@ -0,0 +1,3 @@
+tests:
+  test:
+    tags: cpp
```
if this is going to be a compile test only, i.e. this test is not supposed to run, then please add:

build_only: true

Also, AFAIK c++ is still not supported with native_posix, so you want to add arch_exclude: posix as well.

nashif commented:
@nzmichaelh  please rebase

nzmichaelh commented:
@nashif I've rebased and added the flags to testcase.yaml.  Let's see how the build goes...

nashif commented:
@nzmichaelh 
> @nashif I've rebased and added the flags to testcase.yaml. Let's see how the build goes...

thanks :)
